usr/sbin/p11sak/p11sak.c

```diff
@@ -999,34 +1114,49 @@ CK_BBOOL is_mech_supported(CK_SLOT_ID slot, CK_MECHANISM *pmech,
  */
 static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
                          CK_ULONG keylength, CK_MECHANISM *pmech,
-                         char *attr_string, CK_OBJECT_HANDLE *phkey,
-                         char *label)
+                         char *attr_string, char *custom_attr_string, 
+                         CK_OBJECT_HANDLE *phkey, char *label)
 {
     CK_RV rc;
-    int i = 0;
+    int j, i = 0;
 
     /* Boolean attributes (cannot be specified by user) */
     CK_BBOOL a_token = ckb_true; // always true
 
     /* Non-boolean attributes */
     CK_ULONG a_value_len = keylength / 8;
 
-    /* Standard template */
-    CK_ATTRIBUTE key_attr[3 + KEY_MAX_BOOL_ATTR_COUNT] = {
+    /**Standard template 
+     * 128 is the current count of attributes defined in pkcs11types.h
+     * last updated 08.01.2022
+     */
+    CK_ATTRIBUTE key_attr[128] = {
```

Comment:
Why `128`?

Reply:
see comment above... its the current count of attributes, which are defined in pkcs11types.h and therefore the max length of key_attr

Reply:
Yes, but nothing is guarding against the same attribute multiple times.  So 128 is just an optimistic maximum that can easily be violated...

Reply:
Added Check for multiple attribute definitions

Reply:
please add a define for the number of attributes (also see my comment above)